pkg/agent/puller.go

```diff
@@ -1,73 +1,130 @@
 package agent
 
 import (
-	"fmt"
 	"github.com/kubeflow/kfserving/pkg/agent/storage"
+	v1 "github.com/kubeflow/kfserving/pkg/apis/serving/v1beta1"
 	"log"
 	"path/filepath"
 )
 
+type OpType string
+
+const (
+	Add    OpType = "Add"
+	Remove OpType = "Remove"
+)
+
+// sentinel to represent model absence
+var remove v1.ModelSpec = v1.ModelSpec{}
+
 type Puller struct {
-	ChannelMap map[string]Channel
-	Downloader Downloader
+	channelMap  map[string]ModelChannel
+	completions chan string
```

Comment:
I think it is useful to know the command add or remove on the completion channel(good for me for asserting in tests),  in my PR I also included `error` field to track the download failures but now I think there is a common pattern to have separate error channel. WDYT?

Reply:
Right, I hadn't included proper handling of the error cases in this pass, was thinking that could be done in a next iteration :) Since there's quite a lot to consider and I expect it will require some adjustments of the state management... maybe something like making the per-model goroutines more like reconciliation loops.

Having the completion events be/include the original op for now sounds like a good idea though!

Reply:
I 100% would agree on a seperate error channel. I think that beyond for testing it would be important for the error status endpoint.